runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java

```diff
@@ -185,6 +157,41 @@ public void addNode(Layout currentLayout,
         reconfigureSequencerServers(currentLayout, newLayout, false);
     }
 
+    /**
+     * Heals an existing node in the layout.
+     *
+     * @param currentLayout Current layout.
+     * @param endpoint      New endpoint to be added.
+     * @throws OutrankedException         if consensus outranked.
+     */
+    public void healNode(Layout currentLayout, String endpoint) throws OutrankedException {
+
+        Layout newLayout;
+        if (currentLayout.getUnresponsiveServers().contains(endpoint)) {
+            currentLayout.setRuntime(runtime);
+            sealEpoch(currentLayout);
+
+            LayoutBuilder layoutBuilder = new LayoutBuilder(currentLayout);
+            Layout.LayoutSegment latestSegment =
+                    currentLayout.getSegments().get(currentLayout.getSegments().size() - 1);
+            layoutBuilder.addLogunitServer(0,
+                    getMaxGlobalTail(latestSegment),
+                    endpoint);
+            layoutBuilder.removeUnresponsiveServers(Collections.singleton(endpoint));
+            newLayout = layoutBuilder.build();
+            newLayout.setRuntime(runtime);
+
+            attemptConsensus(newLayout);
+        } else {
+            log.info("Node {} already exists in the layout, skipping.", endpoint);
+            newLayout = currentLayout;
+        }
+
+        // Heal node is successful even if reconfigure sequencer fails.
+        // TODO: Optimize this by retrying or submitting a workflow to retry.
```

Comment:
what are we going to do about this ?

Reply:
We can separate this as another Action in the workflow. Since this is critical to the data path, failure of this will then cause the workflow to fail and will be restarted.